### PR TITLE
Compatibility layer API addition

### PR DESCRIPTION
**Pull request by miyazakh**
Added compatibility layer:

- MD5/MD5_Transform

- SHA/SHA_Transform/SHA1_Transform

- SHA224/SHA256_Transform/SHA512_Transform

- SSL_CTX_get0_param/SSL_CTX_set1_param

- X509_load_crl_file

### CODE REVIEW

File: examples/client/client.c [hunk 1]
```diff
@@ -329,6 +329,29 @@ static void SetKeyShare(WOLFSSL* ssl, int onlyKeyShare, int useX25519,
 #endif
 
 #ifdef WOLFSSL_EARLY_DATA
+static void EarlyDataStatus(WOLFSSL* ssl)
```
Comment (SparkiDev) [inline examples/client/client.c:332]: Suggest moving this to test.h where other common printing functions are.

File: src/ssl.c [hunk 2]
```diff
@@ -16776,7 +16796,28 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
         return 0;
     }
 
+    #if defined(OPENSSL_EXTRA)
+    int wolfSSL_SHA_Transform(WOLFSSL_SHA_CTX* sha,
```
Comment (SparkiDev) [inline src/ssl.c:16800]: I don't see SHA_Transform as an OpenSSL API.
Should be defined to SHA1_Transform if needed.

File: src/ssl.c [hunk 3]
```diff
@@ -42367,6 +42652,16 @@ void* wolfSSL_GetDhAgreeCtx(WOLFSSL* ssl)
     defined(WOLFSSL_HAPROXY)
 
 #ifndef NO_SHA
+    /* One shot SHA hash of message.
+     *
+     * Wrap SHA1 one shot
+     */
+      unsigned char *wolfSSL_SHA(const unsigned char *d, size_t n,
```
Comment (SparkiDev) [inline src/ssl.c:42659]: Make it a define not a function.

File: wolfssl/internal.h [hunk 4]
```diff
@@ -4365,6 +4366,7 @@ struct WOLFSSL {
 #ifdef WOLFSSL_EARLY_DATA
     EarlyDataState earlyData;
     word32 earlyDataSz;
+    word32 earlyDataStatus;
```
Comment (SparkiDev) [inline wolfssl/internal.h:4369]: Change type to byte.

File: examples/client/client.c (hunk 1 above)
Comment (miyazakh) [reply examples/client/client.c]: Thank you for pointing it out. Moved to test.h

File: src/ssl.c (hunk 3 above)
Comment (miyazakh) [reply src/ssl.c]: removed function body and definition, and make  SHA one shot definition to wolfSSL_SHA1

File: wolfssl/internal.h (hunk 4 above)
Comment (miyazakh) [reply wolfssl/internal.h]: Thank you for pointing it out. Changed it to byte.

File: src/ssl.c (hunk 2 above)
Comment (miyazakh) [reply src/ssl.c:16800]: older version of OpenSSL, such as version 1.0.1, has [the API](https://github.com/openssl/openssl/blob/4675a56a3ce2bef715386e45036426c8a4329d39/crypto/sha/sha.h#L115)

Comment (SparkiDev) [reply src/ssl.c:16800]: Thanks!

**Review by SparkiDev — APPROVED** [review]

File: src/ssl.c [hunk 5]
```diff
@@ -16832,6 +16832,26 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
...
+        int ret;
+
+       WOLFSSL_ENTER("MD5_Transform");
+       #if defined(BIG_ENDIAN_ORDER)
+       {
+            ByteReverseWords((word32*)data, (word32*)data, WC_MD5_BLOCK_SIZE);
+       }
+       #endif
+
+       ret = wc_Md5Transform((wc_Md5*)md5, data);
```
Comment (cconlon) [inline src/ssl.c:16847]: It doesn't look like wc_Md5Transform() checks md5 or data against NULL.  We should check/sanitize those arguments coming into wolfSSL_MD5_Transform() first.

File: src/ssl.c [hunk 6]
```diff
@@ -16884,7 +16904,31 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
...
+    {
+       int ret;
+       
+       WOLFSSL_ENTER("SHA_Transform");
+       #if defined(LITTLE_ENDIAN_ORDER)
+       {
+            ByteReverseWords((word32*)data, (word32*)data, WC_SHA_BLOCK_SIZE);
+       }
+       #endif
+       ret = wc_ShaTransform((wc_Sha*)sha, data);
```
Comment (cconlon) [inline src/ssl.c:16921]: It doesn't look like wc_ShaTransform() checks sha or data against NULL.  We should check/sanitize those arguments coming into wolfSSL_SHA_Transform() first.

File: src/ssl.c [hunk 7]
```diff
@@ -17007,8 +17062,32 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
...
+    {
+       int ret;
+       
+       WOLFSSL_ENTER("SHA256_Transform");
+       #if defined(LITTLE_ENDIAN_ORDER)
+       {
+            ByteReverseWords((word32*)data, (word32*)data, WC_SHA256_BLOCK_SIZE);
+       }
+       #endif
+       ret = wc_Sha256Transform((wc_Sha256*)sha256, data);
```
Comment (cconlon) [inline src/ssl.c:17080]: It doesn't look like wc_Sha256Transform() checks sha256 or data against NULL. We should check/sanitize those arguments coming into wolfSSL_SHA256_Transform() first.

File: src/ssl.c [hunk 8]
```diff
@@ -17112,6 +17191,25 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
...
+    #if !defined(HAVE_SELFTEST) && (!defined(HAVE_FIPS) || \
+        (defined(HAVE_FIPS_VERSION) && (HAVE_FIPS_VERSION > 2)))
+    int wolfSSL_SHA512_Transform(WOLFSSL_SHA512_CTX* sha512, 
+                                          const unsigned char* data)
+    {
+       int ret = WOLFSSL_SUCCESS;
+       
+       WOLFSSL_ENTER("SHA512_Transform");
+       
+       ret = wc_Sha512Transform((wc_Sha512*)sha512, data);
```
Comment (cconlon) [inline src/ssl.c:17203]: It doesn't look like wc_Sha512Transform() checks sha512 or data against NULL. We should check/sanitize those arguments coming into wolfSSL_SHA512_Transform() first.

File: src/ssl.c [hunk 9]
```diff
@@ -25980,12 +26078,98 @@ WOLFSSL_X509 *wolfSSL_d2i_X509_fp(XFILE fp, WOLFSSL_X509 **x509)
...
+
+#ifndef NO_BIO
+WOLFSSL_API WOLFSSL_X509_CRL *wolfSSL_d2i_X509_CRL_bio(WOLFSSL_BIO *bp, 
+                                                    WOLFSSL_X509_CRL **x)
+{
+    int derSz;
+    byte* der = NULL;
+    WOLFSSL_X509_CRL* crl = NULL;
+    
+    if ((derSz = wolfSSL_BIO_get_len(bp)) > 0) {
```
Comment (cconlon) [inline src/ssl.c:26090]: Please sanitize "bp" for NULL before using.  It doesn't look like wolfSSL_BIO_get_len() adequately sanitizes.

File: src/ssl.c [hunk 10]
```diff
@@ -25980,12 +26078,98 @@ WOLFSSL_X509 *wolfSSL_d2i_X509_fp(XFILE fp, WOLFSSL_X509 **x509)
...
+        } else {
+            ret = wolfSSL_X509_STORE_add_crl(ctx->store, crl);
+            if (ret == WOLFSSL_FAILURE) {
+                WOLFSSL_MSG("Adding crl failed");
+            } else {
+                ret = 1;/* handled a file */
+            }
+        }
+    } else {
+        WOLFSSL_MSG("Invaid file type");
```
Comment (cconlon) [inline src/ssl.c:26164]: Minor spelling correction - "Invaid" should be "Invalid"

File: src/ssl.c [hunk 11]
```diff
@@ -26195,6 +26379,68 @@ int wolfSSL_X509_VERIFY_PARAM_clear_flags(WOLFSSL_X509_VERIFY_PARAM *param,
...
 
+/*
+* WOLFSSL_VPARAM_DEFAULT          any values in "src" is copied
+*                                 if "src" value is new for "to".
+* WOLFSSL_VPARAM_OVERWRITE        all values of "form" are copied to "to"
+* WOLFSSL_VPARAM_RESET_FLAGS      the flag values are copied, not Ored
+* WOLFSSL_VPARAM_LOCKED           don't copy any values
+* WOLFSSL_VPARAM_ONCE             the current inherit_flags is zerroed
+*/
+static int woflSSL_X509_VERIFY_PARAM_inherit(WOLFSSL_X509_VERIFY_PARAM *to,
```
Comment (cconlon) [inline src/ssl.c:26390]: Spelling - "woflSSL_X509_VERIFY_PARAM_inherit" should have "wolfSSL" spelled correctly - "wolfSSL_X509_VERIFY_PARAM_inherit".

File: src/ssl.c [hunk 12]
```diff
@@ -16832,6 +16832,26 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
 
         return 0;
     }
+
+    int wolfSSL_MD5_Transform(WOLFSSL_MD5_CTX* md5, const unsigned char* data)
```
Comment (cconlon) [inline src/ssl.c:16836]: All new APIs should have at least a one-line comment explaining what the function does.  Please apply this request to all new functions in the PR.  Each new function should have a corresponding descriptive comment.

File: src/ssl.c [hunk 13]
```diff
@@ -26250,6 +26516,32 @@ WOLFSSL_X509_VERIFY_PARAM* wolfSSL_get0_param(WOLFSSL* ssl)
...
+    unsigned int _inherit_flags;
+    
+    if (!to) {
+        return ret;
+    }
+    /* keeps the inherit flags for save */
+    _inherit_flags = to->inherit_flags;
+
+    /* Ored DEFAULT inherit flag proerty to copy "from" contents to "to"
+    *  contends
```
Comment (cconlon) [inline src/ssl.c:26533]: Should "contends" be "contents"?

File: wolfcrypt/src/md5.c [hunk 14]
```diff
@@ -550,7 +550,12 @@ int wc_Md5Copy(wc_Md5* src, wc_Md5* dst)
 
     return ret;
 }
-
+#ifdef OPENSSL_EXTRA
+int wc_Md5Transform(wc_Md5* md5, const byte* data)
+{
+    return Transform(md5, data);
```
Comment (cconlon) [inline wolfcrypt/src/md5.c:556]: Please check/sanitize input arguments

File: wolfcrypt/src/sha.c [hunk 15]
```diff
@@ -773,6 +773,13 @@ int wc_ShaFinal(wc_Sha* sha, byte* hash)
     return ret;
 }
 
+#if defined(OPENSSL_EXTRA)
+int wc_ShaTransform(wc_Sha* sha, const unsigned char* data)
+{
+    return (Transform(sha, data));
```
Comment (cconlon) [inline wolfcrypt/src/sha.c:779]: Please check/sanitize input arguments

File: wolfcrypt/src/sha256.c [hunk 16]
```diff
@@ -1295,6 +1295,12 @@ static int InitSha256(wc_Sha256* sha256)
         return InitSha256(sha256);  /* reset state */
     }
 
+    #if defined(OPENSSL_EXTRA)
+    int wc_Sha256Transform(wc_Sha256* sha, const unsigned char* data)
+    {
+        return (Transform_Sha256(sha, data));
```
Comment (cconlon) [inline wolfcrypt/src/sha256.c:1301]: Please check/sanitize input arguments

File: wolfcrypt/src/sha512.c [hunk 17]
```diff
@@ -916,7 +916,47 @@ void wc_Sha512Free(wc_Sha512* sha512)
     wolfAsync_DevCtxFree(&sha512->asyncDev, WOLFSSL_ASYNC_MARKER_SHA512);
 #endif /* WOLFSSL_ASYNC_CRYPT */
 }
+#if defined(OPENSSL_EXTRA)
+int wc_Sha512Transform(wc_Sha512* sha, const unsigned char* data)
```
Comment (cconlon) [inline wolfcrypt/src/sha512.c:920]: Please check/sanitize input arguments

**Review by cconlon — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 5 above)
Comment (miyazakh) [reply src/ssl.c:16847]: Exactly. Thank you for pointing out!

File: src/ssl.c (hunk 6 above)
Comment (miyazakh) [reply src/ssl.c:16921]: Exactly. Thank you for pointing out!

File: src/ssl.c (hunk 7 above)
Comment (miyazakh) [reply src/ssl.c:17080]: Exactly. Thank you for pointing it out!

File: src/ssl.c (hunk 8 above)
Comment (miyazakh) [reply src/ssl.c:17203]: Exactly. Thank you for pointing it out!

File: src/ssl.c (hunk 13 above)
Comment (miyazakh) [reply src/ssl.c:26533]: Thanks!

File: src/ssl.c (hunk 9 above)
Comment (miyazakh) [reply src/ssl.c:26090]: Agree.

File: wolfcrypt/src/md5.c (hunk 14 above)
Comment (miyazakh) [reply wolfcrypt/src/md5.c:556]: Thanks!

File: wolfcrypt/src/sha.c (hunk 15 above)
Comment (miyazakh) [reply wolfcrypt/src/sha.c:779]: Thanks!

File: wolfcrypt/src/sha256.c (hunk 16 above)
Comment (miyazakh) [reply wolfcrypt/src/sha256.c:1301]: Thanks!

File: wolfcrypt/src/sha512.c (hunk 17 above)
Comment (miyazakh) [reply wolfcrypt/src/sha512.c:920]: Thanks!

File: src/ssl.c (hunk 10 above)
Comment (miyazakh) [reply src/ssl.c:26164]: Thanks!

File: src/ssl.c (hunk 11 above)
Comment (miyazakh) [reply src/ssl.c:26390]: Thanks!

File: src/ssl.c (hunk 12 above)
Comment (miyazakh) [reply src/ssl.c:16836]: Added comments to APIs and functions

**Review by cconlon — APPROVED** [review]

**Pull request merged**